### PR TITLE
Use unified calculate_capacity_factors.

### PR DESCRIPTION
**Pull request by hemna**
This patch adds the new cinder function calculate_capacity_factors.
This method is being added upstream to unify the way cinder calculates
the virtual free space available against a pool.  This also builds
a dictionary of all of the factors that are used to calculate that
virtual free space.   Those factors are also what hammer os poool-stats
reports for each pool(datastore).   Eventually the factors dictionary
will be added to the pools reporting by the scheduler, so cinder will
provide the definitive answer for the factors.  For now, we embed
the new function in hammer.

### CODE REVIEW

**Comment by hemna** [comment]
This patch mimiks the upstream work I have here:

https://review.opendev.org/c/openstack/cinder/+/831247

File: openstack_exporter/collectors/cinderbackend.py [hunk 1]
```diff
@@ -43,40 +43,91 @@ def describe(self):
...
+            free_percent = (virtual_free / total_available_capacity) * 100
+        else:
+            provisioned_ratio = 0
+            free_percent = 0
+
+        return {
+            "total_capacity": total,
+            "free_capacity": free_capacity,
+            "reserved_capacity": reserved_capacity,
+            "total_reserved_available_capacity": int(total - reserved_capacity),
```
Comment (joker-at-work) [inline openstack_exporter/collectors/cinderbackend.py]: Could use `total_avail` here, if you wanted.

File: openstack_exporter/collectors/cinderbackend.py [hunk 2]
```diff
@@ -43,40 +43,91 @@ def describe(self):
...
+            "max_over_subscription_ratio": (
+                max_over_subscription_ratio if provisioned_type == 'thin' else None
```
Comment (joker-at-work) [inline openstack_exporter/collectors/cinderbackend.py:124]: Does it make sense to have `None` here or would `1` maybe serve a better purpose? That way,  one could maybe always use the `max_over_subscription_ratio` in computations, but it wouldn't change the outcome. Depends on how we use it though, so I leave the decision to you.

File: openstack_exporter/collectors/cinderbackend.py [hunk 3]
```diff
@@ -43,40 +43,91 @@ def describe(self):
...
+        (allocated_capacity_gb)
+
+        calculated_free_capacity - total_available_capacity - provisioned_capacity
+        virtual_free_capacity - The calculated free capacity available to cinder
+        to allocate new storage.
+        For thin: calculated_free_capacity
+        For thick: the reported free_capacity can be less than the calculated
+        Capacity, so we use free_capacity - reserved_capacity.
+
+        free_percent - the percentage of the total_available_capacity is left over
```
Comment (joker-at-work) [inline openstack_exporter/collectors/cinderbackend.py]: Needs a better description/name to show whether it's based on `calculated_free_capacity` or `virtual_free_capacity`

**Review by joker-at-work — APPROVED** [review]
I'd probably have indented the lines belonging to the description of the same variable in the function comment to make it a little easier to grasp what belongs together and where I have to jump to to find the next variable.

All-in-all, I couldn't find anything bad, just some things you might want to take a look at.

File: openstack_exporter/collectors/cinderbackend.py (hunk 2 above)
Comment (hemna) [reply openstack_exporter/collectors/cinderbackend.py:124]: I've gone back and forth on this too.   I had set it to 1 at one point, but if the provisioning type is thick max_over_subscription_ratio should never be used.   So, that's why I settled on None.

File: openstack_exporter/collectors/cinderbackend.py (hunk 1 above)
Comment (hemna) [reply openstack_exporter/collectors/cinderbackend.py]: good idea, I'll update this

File: openstack_exporter/collectors/cinderbackend.py (hunk 3 above)
Comment (hemna) [reply openstack_exporter/collectors/cinderbackend.py]: done

**Pull request merged**